---
title: "figure 2 joint for Philip and Calderon data"
author: "Koen Van den Berge"
date: "2/26/2020"
output: html_document
---

# Load Calderon

```{r}
library(GenomicAlignments)
library(tidyverse)
library(cqn)
library(RColorBrewer)
library(edgeR)
library(ggplot2)
library(cowplot)
library(ggridges)
library(gridExtra)
colsBig <- clusterExperiment:::massivePalette
plotGCHex <- function(gr, counts){
  counts2 <- counts
  df <- as_tibble(cbind(counts2,gc=mcols(gr)$gc))
  df <- gather(df, sample, value, -gc)
  ggplot(data=df, aes(x=gc, y=log(value+1)) ) + 
    ylab("log(count + 1)") + xlab("GC-content") + 
    geom_hex(bins = 50) + theme_bw() #+ facet_wrap(~sample, nrow=2)
}
pal <- RColorBrewer::brewer.pal(n=8, "Dark2")
source("../../methods/gcqn_validated.R")

### this dataset combines samples from a number of different sources, therefore the GC effects are wildly different between samples.
data=read.delim("../../data/calderon2019_GSE118189/GSE118189_ATAC_counts.txt.gz")
colnames(data) <- substr(colnames(data),2,nchar(colnames(data)))

# get GC content
rn <- rownames(data)
sn <- unlist(lapply(lapply(strsplit(rn,split="_"),"[",1),function(x) gsub(pattern="chr",x=x,replacement="")))
start <- as.numeric(unlist(lapply(strsplit(rn,split="_"),"[",2)))
end <- as.numeric(unlist(lapply(strsplit(rn,split="_"),"[",3)))
gr <- GRanges(seqnames=sn, ranges=IRanges(start, end), strand="*", mcols=data.frame(peakID=rn))
ff <- FaFile("~/data/genomes/human/Homo_sapiens.GRCh37.75.dna.primary_assembly.fa.gz")
peakSeqs <- getSeq(x=ff, gr)
gcContentPeaks <- letterFrequency(peakSeqs, "GC",as.prob=TRUE)[,1]
gcGroups <- Hmisc::cut2(gcContentPeaks, g=20)
mcols(gr)$gc <- gcContentPeaks

# get metadata
cnames <- colnames(data)
donor <- unlist(lapply(strsplit(cnames,split=".",fixed=TRUE),"[",1))
celltype <- factor(unlist(lapply(strsplit(cnames,split=".",fixed=TRUE),"[",2)))
condition <- unlist(lapply(strsplit(cnames,split=".",fixed=TRUE),"[",3))
table(celltype,condition)

# QC measures
qcMeasures <- openxlsx::read.xlsx(xlsxFile = "../../data/calderon2019_GSE118189/Supplementary_tables.xlsx",
                    sheet = 5)
qcMeasures <- qcMeasures[,1:4]
qcMeasures$sample <- gsub(x=qcMeasures$sample, pattern="-", replacement=".", fixed=TRUE)
rownames(qcMeasures) <- qcMeasures$sample
qcMeasures <- qcMeasures[colnames(data),]

# batch
addSamples <- read.table("../../data/calderon2019_GSE118189/samples_with_additional_resequencing.txt",
                         stringsAsFactors = FALSE)[,1]
addSamples <- gsub(x=addSamples, pattern="-", replacement=".", fixed=TRUE)
batch2 <- rep(0, ncol(data))
names(batch2) <- colnames(data)
batch2[addSamples]<- 1
batch2 <- factor(batch2)
batch <- droplevels(interaction(donor, batch2))

table(celltype,condition, batch)
counts <- as.matrix(data) ; rm(data) ; gc()
```


## Mock comparisons

```{r}
# Memory_NK cells
memContID <- celltype == "Memory_NK" & condition == "U"
countsMemControl <- counts[,memContID]
keepMemContr <- rowSums(cpm(countsMemControl) >= 2) >=3 
countsMemControl <- countsMemControl[keepMemContr, ]
# these are all from different batches.
table(droplevels(batch[memContID]))
# equally sized bins after filtering
gcGroupsMem <- Hmisc::cut2(gcContentPeaks[keepMemContr], g=20)
gcGroupsMem10 <- Hmisc::cut2(gcContentPeaks[keepMemContr], g=10)
gcMem <- gcContentPeaks[keepMemContr]

set.seed(33)
mock <- factor(sample(rep(letters[1:2], each=3)))
design <- model.matrix(~mock)
```

```{r}

##### FIGURE 2
## cqn
cqnModel <- cqn(countsMemControl, x=gcMem, sizeFactors = colSums(countsMemControl),
                lengths=width(gr)[keepMemContr])
d <- DGEList(countsMemControl)
d$offset <- cqnModel$glm.offset
d <- estimateDisp(d, design)
fit <- glmFit(d, design)
lrtCqn <- glmLRT(fit, coef=2)
dfCqn <- data.frame(logFC=log(2^lrtCqn$table$logFC),
                   gc=gcGroupsMem)
pCqn <- ggplot(dfCqn) +
  aes(x=gc, y=logFC, color=gc) +
  geom_violin() +
  geom_boxplot(width=0.1) +
  scale_color_manual(values=wesanderson::wes_palette("Zissou1", nlevels(gcGroups), "continuous")) +
  geom_abline(intercept = 0, slope = 0, col="black", lty=2) +
  theme_bw() + 
  ylim(c(-1,1)) +
  xlab("GC-content bin") +
  ggtitle("cqn normalization") +
  theme(axis.text.x = element_text(angle = 45, vjust = .5),
        legend.position = "none",
        axis.title = element_text(size=16)) +
  geom_smooth(se=FALSE, color="blue", aes(group=1), lwd=1)

# ## EDASeq
library(EDASeq)
#emptyRows <- which(rownames(countsMouse) == "")
#rownames(countsMouse)[emptyRows] <- paste0("emptyPeak",1:length(emptyRows))
dataWithin <- withinLaneNormalization(countsMemControl, y=gcMem,
                                      num.bins=20, which="full")
dataNorm <- betweenLaneNormalization(dataWithin, which="full")
d <- DGEList(dataNorm)
d <- estimateDisp(d, design)
fit <- glmFit(d, design)
lrtEDASeq <- glmLRT(fit, coef=2)
dfEDASeq <- data.frame(logFC=log(2^lrtEDASeq$table$logFC),
                    gc=gcGroupsMem)
pEDASeq <- ggplot(dfEDASeq) +
  aes(x=gc, y=logFC, color=gc) +
  geom_violin() +
  geom_boxplot(width=0.1) +
  scale_color_manual(values=wesanderson::wes_palette("Zissou1", nlevels(gcGroups), "continuous")) +
  geom_abline(intercept = 0, slope = 0, col="black", lty=2) +
  theme_bw() +
  ylim(c(-1,1)) +
  xlab("GC-content bin") +
  ggtitle("FQ-FQ normalization") +
  theme(axis.text.x = element_text(angle = 45, vjust = .5),
        legend.position = "none",
        axis.title = element_text(size=16)) +
  geom_smooth(se=FALSE, color="blue", aes(group=1), lwd=1)


## GC-QN
countsGCQN <- gcqn(countsMemControl, gcGroupsMem, summary = "median")
d <- DGEList(countsGCQN)
d <- estimateDisp(d, design)
fit <- glmFit(d, design)
lrtGCQN <- glmLRT(fit, coef=2)
dfGCQN <- data.frame(logFC=log(2^lrtGCQN$table$logFC),
                   gc=gcGroupsMem)
pGCQN <- ggplot(dfGCQN) +
  aes(x=gc, y=logFC, color=gc) +
  geom_violin() +
  geom_boxplot(width=0.1) +
  scale_color_manual(values=wesanderson::wes_palette("Zissou1", nlevels(gcGroups), "continuous")) +
  geom_abline(intercept = 0, slope = 0, col="black", lty=2) +
  theme_bw() + 
  ylim(c(-1,1)) +
  xlab("GC-content bin") +
  ggtitle("GC-FQ normalization") +
  theme(axis.text.x = element_text(angle = 45, vjust = .5),
        legend.position = "none",
        axis.title = element_text(size=16)) +
  geom_smooth(se=FALSE, color="blue", aes(group=1), lwd=1)
```


# Load Philip data

```{r}
files <- list.files("../../data/philip2017_GSE89308/data", full=TRUE)
countsAll <- sapply(files, function(x) read.delim(x), simplify=FALSE)

counts1 <- countsAll[[1]]
regionsAllGR <- GRanges(seqnames=unlist(lapply(strsplit(as.character(counts1$chr),split="chr"), function(x) x[2])), ranges=IRanges(counts1$start, counts1$end), strand="*", mcols=data.frame(symbol=counts1$symbol, counts=counts1$N1))
ff <- FaFile("~/data/genomes/mouse/Mus_musculus.GRCm38.dna.primary_assembly.fa")
peakSeqs <- getSeq(x=ff, regionsAllGR)
gcContentPeaks <- letterFrequency(peakSeqs, "GC",as.prob=TRUE)[,1]
gcGroups <- Hmisc::cut2(gcContentPeaks, g=20)
mcols(regionsAllGR)$gc <- gcContentPeaks


humanSamples <- 48:length(countsAll)
mouseSamples <- 1:41

countsMouse <- do.call(cbind,lapply(countsAll[mouseSamples], function(x) x[,7]))
colnames(countsMouse) <- unlist(lapply(strsplit(unlist(lapply(strsplit(names(countsAll[mouseSamples]), split="/"),"[[",6)),split="_"), function(x) paste(x[1:3],collapse="_")))
rownames(countsMouse) <- mcols(regionsAllGR)$mcols.symbol
sf <- readRDS("../../data/philip2017_GSE89308/sf.rds")
countsMouse <- round(t(t(countsMouse)*sf))

condition <- unlist(lapply(strsplit(colnames(countsMouse),split="_"),"[[",3))
# fix condition vector
condition[condition%in%paste0("N",1:3)]="N"
condition[condition%in%paste0("M",1:3)]="M"
condition <- factor(condition)
```

```{r}
## ridges plot 
countsN <- countsMouse[,condition == "L5"][,1:3]
gcGroups10 <- Hmisc::cut2(gcContentPeaks, g=10)

lc <- log1p(c(countsN))
joyDat <- data.frame(lc=lc, 
                     gc=rep(gcGroups10, 3),
                     sample=rep(1:3, each=nrow(countsN)))
axText <- 0
pRidge1 <- joyDat %>% ggplot(aes(y=gc)) + 
  geom_density_ridges(aes(x=lc)) + 
  facet_wrap(.~sample, nrow=1) +
  theme_ridges(grid=FALSE, font_size=5, center_axis_labels = TRUE) + 
  xlim(c(1.5,9)) +
  xlab("log(count + 1)") +
  ylab("GC-content bin") +
  theme(axis.text.y = element_text(size=axText),
        axis.text.x = element_text(size=10),
        legend.position = "none",
        axis.title = element_text(size=16), 
             strip.background = element_blank(),
             strip.text.x = element_blank())
```



```{r}
pFC <- cowplot::plot_grid(pCqn, pEDASeq, pGCQN,
                          labels=letters[2:4],
                          nrow=3, ncol=1)

pFig2 <- cowplot::plot_grid(pRidge1, 
                   pFC,
                   labels=c("a",""))
pFig2
ggsave("~/Dropbox/research/atacseq/bulk/plots/figure2_joint.png",
       units="in", width=12, height=9)
```


