---
title: "Fullard 2019 ATAC-seq human brain atlas"
author: "Koen Van den Berge"
date: "11/27/2019"
output: html_document
---


```{r}
library(GenomicAlignments)
library(edgeR)
library(UpSetR)
library(rafalib)
library(tidyverse)
library(ggridges)
library(tidyr)
library(quantro)
library(qsmooth)
library(RUVSeq)
library(scone)
library(rhdf5)

ridgesPlot <- function(counts, gcGroups){
  #logcounts <- log1p(counts)
  logcounts <- as.data.frame(edgeR::cpm(counts, log=TRUE))
  logcountsLong <- gather(logcounts)
  dat <- data.frame(logcountsLong, gc=rep(gcGroups, ncol(counts)))
  print(dat %>% ggplot(aes(y=gc)) + geom_density_ridges(aes(x=value, fill=key, alpha=.2)) + theme_ridges(grid=FALSE))
}

runEdgeR <- function(counts, design, tmm=TRUE, offset=NULL){
  d <- DGEList(counts)
  if(!is.null(offset)) d$offset <- offset
  if(tmm){
      d <- calcNormFactors(d) 
    } else {
      d$samples$lib.size = 1
  }
  d <- estimateDisp(d,design)
  fit <- glmFit(d, design)
  return(fit)
}
fcBoxplots <- function(fit, gcGroups){
  mypar(mfrow=c(3,2))
  for(ii in 2:ncol(fit$coefficients)){
    boxplot(fit$coefficients[,ii]~gcGroups,
         ylim=c(-1,1)) ; abline(h=0, col="red", lwd=2)
    df <- data.frame(beta=fit$coefficients[,ii], gcl=gcGroups)
  hlp <- df %>% group_by(gcl) %>% summarize(betaMean=mean(beta))
  lines(x=1:nlevels(gcGroups), y=hlp$betaMean,
        col="orange" , type="b", lwd=2)
  }
  mypar(mfrow=c(1,1))
}

plotCountMatrices <- function(mat1, mat2, gcGroups, ...){
  pal <- wesanderson::wes_palette("Zissou1", n=20, type="continuous")
  plot(x=unlist(c(mat1+1)), 
     y=unlist(c(mat2+1)),
     pch=16, cex=1/2,
     col=pal[gcGroups],
     ...) ; 
  abline(0,1, col="red")
}


counts <- read.table("~/data/atac/fullard2019/boca_raw_count_matrix.tsv", 
                     header=TRUE, stringsAsFactors = FALSE)

```

```{r}
peaks <- read.table("~/data/atac/fullard2019/boca_peaks_consensus_no_blacklisted_regions.bed", header=FALSE, stringsAsFactors = FALSE)
colnames(peaks) <- c("chromosome", "start", "end", "name")

peakNames <- peaks$name
sn <- substr(peaks$chromosome, 4, sapply(peaks$chromosome, nchar))
start <- peaks$start
end <- peaks$end
gr <- GRanges(seqnames=sn, ranges = IRanges(start=start, end=end), strand="*")
names(gr) <- peaks$name

# get GC content
ff <- FaFile("~/data/genomes/human/Homo_sapiens.GRCh37.75.dna_sm.primary_assembly.fa")
peakSeqs <- getSeq(x=ff, gr)
gcContentPeaks <- letterFrequency(peakSeqs, "GC",as.prob=TRUE)[,1]
gcContentPeaks <- sample(gcContentPeaks)
gcGroups <- Hmisc::cut2(gcContentPeaks, g=20)
mcols(gr)$gc <- gcContentPeaks

# design
# the data should consist of 2 cell types (neurons and non-neurons) across 14 distinct brain regions of 5 individuals
colnames(counts) <- gsub(x=colnames(counts),pattern="^X", replacement="")
individual <- factor(unlist(lapply(strsplit(colnames(counts), split="_"), "[[", 1)))
celltype <- factor(unlist(lapply(strsplit(colnames(counts), split="_"), "[[", 2)))
region <- factor(unlist(lapply(strsplit(colnames(counts), split="_"), "[[", 3)))

# metadata
meta <- readr::read_delim("~/data/atac/fullard2019/metadata_boca.csv", 
                   delim="\t")
table(meta$Sequence_date, meta$Flowcell)
table(meta$Sequence_date, meta$Sort_date)

```


```{r}
bpparam <- BiocParallel::bpparam()
bpparam$workers <- 1

source("~/Dropbox/research/atacseq/method/scone_evaluation.R")

###  filter 
keep <- rowSums(edgeR::cpm(counts) > 2) >=10
counts <- counts[keep,]
gcContentPeaks <- gcContentPeaks[keep]
gcGroups <- gcGroups[keep]
gr <- gr[keep]

expr <- as.matrix(counts)
bio <- droplevels(interaction(celltype, region))
hk <- readRDS("~/data/genomes/hkListHumanGenomicRanges.rds")
# get peaks that are within hk
qh <- queryHits(findOverlaps(gr, hk, type="within"))
negcon <- rownames(counts)[qh]
# the Sort_date inlcudes sequence_date so I'll consider this as batch
# note that other batch effects may be present, e.g. PCR date.
batch <- factor(meta$Sort_date) 
qc <- data.frame(meta$Input_mg, meta$Total_PCR_cycles, meta$Gel_date)
qc <- cbind(qc, meta[,52:ncol(meta)])
# remove constant variables or variables about biology of the expt.
qc <- qc[,!colnames(qc) %in% c("star_Insertion_rate_per_base", "star_Deletion_average_length", "picard_SECONDARY_OR_SUPPLEMENTARY_RDS", "picard_UNMAPPED_READS", "metaMergingDesigns", "lowReadCount", "cell_subtype", "readCountInfo_Unassigned_MultiMapping", "readCountInfo_Unassigned_Unmapped", "readCountInfo_Unassigned_MappingQuality", "readCountInfo_Unassigned_Chimera", "readCountInfo_Unassigned_Secondary")]
qc2 <- apply(qc,2,as.numeric)
# replace pctages with numerics
pctVars <- grep(x=colnames(qc2), pattern="_pct")
qcPct <- apply(qc[,pctVars], 2, function(var){
  as.numeric(gsub(x=var, pattern = "%", replacement=""))
})
qc2[,pctVars] <- qcPct
# remove others
fctID <- which(apply(qc2,2,function(var) any(is.na(var))))
qc2 <- qc2[,-fctID]



gcContentScone <- gcContentPeaks
gcGroupsScone <- gcGroups
h5File <- "/Users/koenvandenberge/tmp/normCounts_fullard2019_permuted.h5"
# different number of bins
gcGroupsScone10 <- Hmisc::cut2(gcContentPeaks, g=10)
gcGroupsScone50 <- Hmisc::cut2(gcContentPeaks, g=50)
gcGroupsScone100 <- Hmisc::cut2(gcContentPeaks, g=100)
sconeFit <- sconeEvaluation(counts=counts, bio=bio, batch=batch, qc=qc2,
                            negcon=negcon, gcContent=gcContentPeaks, 
                            gcGroups=gcGroups, k_ruv=5, k_qc=2, 
                            adjust_bio='no', eval_kclust = nlevels(bio),
                            diffLength=TRUE, bpparam=bpparam, return_norm="hdf5",
                            hdf5file=h5File)

## add GC-bias through RLE to scores
rleGC_med <- rleGCh5(h5File = h5File, 
                  gcContent = gcContentScone,
                  type="med")
rleGC_iqr <- rleGCh5(h5File = h5File, 
                  gcContent = gcContentScone,
                  type="iqr")
scores <- get_scores(sconeFit)
# negative is good.
scores <- cbind(scores, 
                rleGC_med = -rleGC_med[rownames(scores)],
                rleGC_iqr = -rleGC_iqr[rownames(scores)])
## recalculate score ranks with new metric
scoreT <- t(scores)
ranked_scores <- apply(scoreT,1,rank)
mean_score_rank <- rowMeans(ranked_scores)
mean_score_rank <- mean_score_rank[order(mean_score_rank, decreasing=TRUE)]
mean_score_rank

```

# Biplot

```{r}
pc_obj <- prcomp(apply(scoreT,1,rank),
                center = TRUE,scale = FALSE)
methodLabelsTmp <- strsplit(rownames(pc_obj$x), split=",")
methodNames <- unlist(lapply(methodLabelsTmp, function(x){
  paste0(x[2:3], collapse=",")
  #x[2]
}))
methods <- unlist(lapply(strsplit(methodNames, split=","), "[[", 1))
ruvs <- unlist(lapply(strsplit(methodNames, split=","), function(x) strsplit(x[2],split="_")))[seq(2,length(methodNames)*2, by=2)]
methodNames <- paste(methods,ruvs, sep = "_")

bp_obj <- biplot_color(pc_obj,y = -mean_score_rank,expand = .6, cex=2)
text(x=bp_obj[,1], y=bp_obj[,2], labels=methodNames, cex=2/3)
```

# RLE Before and after

```{r}
meds <- rowMedians(as.matrix(counts))
rleIk <- log(sweep(as.matrix(counts),1,meds,FUN="/"))
boxplot(rleIk, ylab="RLE", xlab="Sample", main="RLE before normalization")
abline(h=0, col="red", lty=2, lwd=3/2)

out_norm = get_normalized(sconeFit,
                          method = rownames(get_params(sconeFit))[1])
meds <- rowMedians(as.matrix(out_norm))
rleIk <- log(sweep(as.matrix(out_norm),1,meds,FUN="/"))
boxplot(rleIk, ylab="RLE", xlab="Sample", main="RLE after normalization")
abline(h=0, col="red", lty=2, lwd=3/2)
```



```{r}
file.remove(h5File)
saveRDS(scores, file="~/Dropbox/research/atacseq/bulk/analysis/benchmarkNormalization/permuted/fullard2019_scores_permuted.rds")
sessionInfo()
```

