---
title: 'Normalization benchmark: Schizophrenia dataset'
author: "Koen Van den Berge"
date: "12/11/2019"
output: html_document
---

```{r}
library(edgeR) ; library(cqn) ; library(GenomicAlignments)
library(scone)
library(qsmooth)
library(rhdf5)

data=read.delim("../../data/bryois2018_schizophrenia/quant_peaks_1FDR_minOverlap2_300bpExt_300bp_centred_raw_288.txt.gz")
counts <- as.matrix(data[,-c(1:4)])
rownames(counts) <- data$ID

# get GC content
sn <- gsub(pattern="chr",x=data[,1],replacement="")
start <- data[,2]
end <- data[,3]
gr <- GRanges(seqnames=sn, ranges=IRanges(start, end), strand="*", mcols=data.frame(peakID=data[,4]))
ff <- FaFile("~/data/genomes/human/Homo_sapiens.GRCh37.75.dna_sm.primary_assembly.fa")
peakSeqs <- getSeq(x=ff, gr)
gcContentPeaks <- letterFrequency(peakSeqs, "GC",as.prob=TRUE)[,1]
gcGroups <- Hmisc::cut2(gcContentPeaks, g=20)

# get metadata
meta <- read.table("../../data/bryois2018_schizophrenia/tableS9.txt", header=TRUE)
cnames <- colnames(counts)
condition <- meta$Dx[match(meta$Sample, cnames)]
table(condition)

rmSamples <- which(condition %in% c("AFF", "BP"))
condition <- condition[-rmSamples]
counts <- counts[,-rmSamples]

rm(data, start, end, peakSeqs) ; gc()
```

# Benchmark normalization

```{r}
bpparam <- BiocParallel::bpparam()
bpparam$workers <- 1


source("../../methods/scone_evaluation.R")
hk <- readRDS("../../objects/hkListHumanGenomicRanges.rds")
qh <- queryHits(findOverlaps(gr, hk, type="within"))
negcon <- rownames(counts)[qh]
length(negcon)
bio <- droplevels(condition)
qcVars <- c("PCR_cycles", "Processed.reads", "unique_align.reads", 
            "reads.aligned.to.chrM", "Gender", 
            "Ethnicity", "Age_of_Death", "Year_of_Autopsy", "Benzodiazepines",
            "Anticonvulsants", "AntipsychTyp", "AntipsychAtyp", "Antidepress",
            "Lithium", "Alcohol", "Illicitsub",   "CauseOfDeath",
            "Seizures", "Braak_Stage", "H.O_COPD", "H.O_Stroke",
            "Lobotomy", paste0("PC",1:10,"_genotype"))
qc <- meta[-rmSamples,qcVars]
batch <- factor(meta[-rmSamples,"sequencing.Index"])

gcContentScone <- gcContentPeaks
gcGroupsScone <- gcGroups
# different number of bins
gcGroupsScone10 <- Hmisc::cut2(gcContentPeaks, g=10)
gcGroupsScone50 <- Hmisc::cut2(gcContentPeaks, g=50)
gcGroupsScone100 <- Hmisc::cut2(gcContentPeaks, g=100)
h5File <- "/Users/koenvandenberge/tmp/normCounts_bryois2018.h5"
sconeFit <- sconeEvaluation(counts=counts, bio=bio, negcon=negcon, qc=qc, batch=batch,
                gcContent=gcContentPeaks, gcGroups=gcGroups, k_ruv=5,
                k_qc=5, adjust_bio='no', eval_kclust = nlevels(bio),
                diffLength=FALSE, bpparam=bpparam, return_norm="hdf5",
                hdf5file=h5File)
saveRDS(sconeFit, "~/Dropbox/research/atacseq/bulk/analysis/benchmarkNormalization/bryois2018_sconeFit.rds")

## add GC-bias through RLE to scores
rleGC_med <- rleGCh5(h5File = h5File, 
                  gcContent = gcContentScone,
                  type="med")
rleGC_iqr <- rleGCh5(h5File = h5File, 
                  gcContent = gcContentScone,
                  type="iqr")
scores <- get_scores(sconeFit)
scores <- cbind(scores, 
                rleGC_med = -rleGC_med[rownames(scores)],
                rleGC_iqr = -rleGC_iqr[rownames(scores)])
## recalculate score ranks with new metrics
scoreT <- t(scores)
ranked_scores <- apply(scoreT,1,rank)
mean_score_rank <- rowMeans(ranked_scores)
mean_score_rank <- mean_score_rank[order(mean_score_rank, decreasing=TRUE)]
mean_score_rank

rm(hk, qh) ; gc()
```


# Biplot

```{r}
pc_obj <- prcomp(apply(scoreT,1,rank),
                center = TRUE,scale = FALSE)
methodLabelsTmp <- strsplit(rownames(pc_obj$x), split=",")
methodNames <- unlist(lapply(methodLabelsTmp, function(x){
  paste0(x[2:3], collapse=",")
}))
methods <- unlist(lapply(strsplit(methodNames, split=","), "[[", 1))
ruvs <- unlist(lapply(strsplit(methodNames, split=","), function(x) strsplit(x[2],split="_")))[seq(2,length(methodNames)*2, by=2)]
methodNames <- paste(methods,ruvs, sep = "_")

bp_obj <- biplot_color(pc_obj,y = -mean_score_rank,expand = .6, cex=2)
text(x=bp_obj[,1], y=bp_obj[,2], labels=methodNames, cex=2/3)
```


```{r}
file.remove(h5File)
saveRDS(scores, file="~/Dropbox/research/atacseq/bulk/analysis/benchmarkNormalization/bryois2018_scores.rds")
sessionInfo()
```

