---
title: "Liu (2019) simulation"
author: "Koen Van den Berge"
date: "3/9/2021"
output: html_document
---

```{r}
library(GenomicAlignments)
library(edgeR)
library(UpSetR)
library(rafalib)
library(tidyverse)
library(ggridges)
library(tidyr)
library(quantro)
library(qsmooth)
library(RUVSeq)
library(scone)
library(rhdf5)
library(cqn)
source("../../methods/gcqn_validated.R")
gcqn_qsmooth <- function(counts, gcGroups, bio){
  gcBinNormCounts <- matrix(NA, nrow=nrow(counts), ncol=ncol(counts), dimnames=list(rownames(counts),colnames(counts)))
  for(ii in 1:nlevels(gcGroups)){
    id <- which(gcGroups==levels(gcGroups)[ii])
    countBin <- counts[id,]
    qs <- qsmooth(countBin, group_factor=bio)
    normCountBin <- qs@qsmoothData
    normCountBin <- round(normCountBin)
    normCountBin[normCountBin<0] <- 0
    gcBinNormCounts[id,] <- normCountBin
  }
  return(gcBinNormCounts)
}
source("../../methods/simulation.R")

counts <- read.table("../../data/liu2019_atlas/ATACseq_Matrix/chromatin.accessibility.raw.count.txt", header=TRUE, stringsAsFactors = FALSE)

```

```{r}
peakNames <- as.character(counts[,1])
peakNamesSplit <- strsplit(peakNames, split="_")
# remove peaks in random chromosomes
rmPeaks <- which(lapply(peakNamesSplit, length) == 4)
counts <- counts[-rmPeaks,]
peakNamesSplit <- peakNamesSplit[-rmPeaks]
sn <- unlist(lapply(peakNamesSplit, "[[", 1))
sn <- substr(sn, 4, nchar(sn))
start <- as.numeric(unlist(lapply(peakNamesSplit, "[[", 2)))
end <- as.numeric(unlist(lapply(peakNamesSplit, "[[", 3)))
gr <- GRanges(seqnames=sn, ranges = IRanges(start=start, end=end), strand="*")
names(gr) <- paste0("peak", 1:length(gr))


# get GC content
ff <- FaFile("~/data/genomes/mouse/Mus_musculus.NCBIM37.67.dna.toplevel.fa")
peakSeqs <- getSeq(x=ff, gr)
gcContentPeaks <- letterFrequency(peakSeqs, "GC",as.prob=TRUE)[,1]
gcGroups <- Hmisc::cut2(gcContentPeaks, g=20)
mcols(gr)$gc <- gcContentPeaks

# remove peak ranges
counts <- counts[,-1]
# restrict to samples discussed in paper
counts <- counts[,!substr(colnames(counts),1,2) == "D0"]

# design
gender <- factor(unlist(lapply(strsplit(colnames(counts), split="_"), "[[", 1)))
tissue <- factor(unlist(lapply(strsplit(colnames(counts), split="_"), "[[", 2)))

# metadata
meta <- openxlsx::read.xlsx("../../data/liu2019_atlas/Metadata.xlsx")
meta <- meta[-nrow(meta),]
metaNames <- paste(meta$Gender, gsub(x=meta$Tissue, pattern=" ", replacement="_"), paste0("Rep",meta$Replicate), sep = "_") # same format as sample names
# same order as count matrix
meta <- meta[match(colnames(counts), metaNames),]
```

# Mock comparison


```{r}
bio <- droplevels(interaction(gender,tissue))
mockGroup <- splitSamples(bio)
table(bio, mockGroup)

keepSamplesMock <- which(!is.na(mockGroup))
countsMock <- counts[,keepSamplesMock]
mockGroup <- mockGroup[keepSamplesMock]

mockEvalRes <- mockEvaluation(simCounts = countsMock,
                               grpSim = mockGroup,
                               simGC = gcContentPeaks,
                               simWidth = width(gr),
                               grSim = gr)
```

# Simulate data to add signal

```{r}
signalRes <- signalEvaluation(counts = countsMock,
                              grp = mockGroup,
                               gc = gcContentPeaks,
                               gr = gr,
                               nTotal = 12)
```


# Gather all results

```{r}
resultList <- list(mockEvalRes, signalRes)
saveRDS(resultList, file="../../objects/liu2019_simulationResults.rds")
```


# Session info

```{r}
sessionInfo()
```



