---
title: "Fullard 2019 ATAC-seq human brain atlas normalizatoin benchmark"
author: "Koen Van den Berge"
date: "11/27/2019"
output: html_document
---


```{r}
library(GenomicAlignments)
library(edgeR)
library(UpSetR)
library(rafalib)
library(tidyverse)
library(ggridges)
library(tidyr)
library(quantro)
library(qsmooth)
library(RUVSeq)
library(scone)
library(rhdf5)
library(cqn)
source("../../methods/gcqn_validated.R")
gcqn_qsmooth <- function(counts, gcGroups, bio){
  gcBinNormCounts <- matrix(NA, nrow=nrow(counts), ncol=ncol(counts), dimnames=list(rownames(counts),colnames(counts)))
  for(ii in 1:nlevels(gcGroups)){
    id <- which(gcGroups==levels(gcGroups)[ii])
    countBin <- counts[id,]
    qs <- qsmooth(countBin, group_factor=bio)
    normCountBin <- qs@qsmoothData
    normCountBin <- round(normCountBin)
    normCountBin[normCountBin<0] <- 0
    gcBinNormCounts[id,] <- normCountBin
  }
  return(gcBinNormCounts)
}
source("../../methods/simulation.R")



counts <- read.table("../../data/fullard2019/boca_raw_count_matrix.tsv", 
                     header=TRUE, stringsAsFactors = FALSE)

```

```{r}
peaks <- read.table("../../data/fullard2019/boca_peaks_consensus_no_blacklisted_regions.bed", header=FALSE, stringsAsFactors = FALSE)
colnames(peaks) <- c("chromosome", "start", "end", "name")

peakNames <- peaks$name
sn <- substr(peaks$chromosome, 4, sapply(peaks$chromosome, nchar))
start <- peaks$start
end <- peaks$end
gr <- GRanges(seqnames=sn, ranges = IRanges(start=start, end=end), strand="*")
names(gr) <- peaks$name

# get GC content
ff <- FaFile("~/data/genomes/human/Homo_sapiens.GRCh37.75.dna_sm.primary_assembly.fa.gz")
peakSeqs <- getSeq(x=ff, gr)
gcContentPeaks <- letterFrequency(peakSeqs, "GC",as.prob=TRUE)[,1]
gcGroups <- Hmisc::cut2(gcContentPeaks, g=20)
mcols(gr)$gc <- gcContentPeaks

# design
# the data should consist of 2 cell types (neurons and non-neurons) across 14 distinct brain regions of 5 individuals
colnames(counts) <- gsub(x=colnames(counts),pattern="^X", replacement="")
individual <- factor(unlist(lapply(strsplit(colnames(counts), split="_"), "[[", 1)))
celltype <- factor(unlist(lapply(strsplit(colnames(counts), split="_"), "[[", 2)))
region <- factor(unlist(lapply(strsplit(colnames(counts), split="_"), "[[", 3)))

# metadata
meta <- readr::read_delim("../../data/fullard2019/metadata_boca.csv", 
                   delim="\t")
table(meta$Sequence_date, meta$Flowcell)
table(meta$Sequence_date, meta$Sort_date)

###  filter 
keep <- rowSums(edgeR::cpm(counts) > 2) >=10
counts <- counts[keep,]
gcContentPeaks <- gcContentPeaks[keep]
gcGroups <- gcGroups[keep]
gr <- gr[keep]
```

# Mock comparison

```{r}
bio <- droplevels(interaction(celltype, region))
mockGroup <- splitSamples(bio)
table(bio, mockGroup)

keepSamplesMock <- which(!is.na(mockGroup))
countsMock <- counts[,keepSamplesMock]
mockGroup <- mockGroup[keepSamplesMock]


mockEvalRes <- mockEvaluation(simCounts = countsMock,
                               grpSim = mockGroup,
                               simGC = gcContentPeaks,
                               simWidth = width(gr),
                               grSim = gr)


```



# Simulate data to add signal

```{r}
signalRes <- signalEvaluation(counts = countsMock,
                              grp = mockGroup,
                               gc = gcContentPeaks,
                               gr = gr,
                               nTotal = 12)
simData <- signalRes$simData
saveRDS(simData, "../../data/simulatedData/fullard2019SimData.rds")
signalRes <- signalRes[-which(names(signalRes) %in% "simData")]
```





# Gather all results

```{r}
resultList <- list(mockEvalRes, signalRes)
saveRDS(resultList, file="../../objects/fullard2019_simulationResults.rds")
```
